### PR TITLE
Composite monitors

### LINKED ISSUE #610: [FEATURE] Composite monitors
**Is your feature request related to a problem?**
Details provided in the [RFC](https://github.com/opensearch-project/alerting/issues/697)

**What solution would you like?**
Create UX for visually performing CRUD operations on composite monitors as well as help with investigations when alerts are created.

**What alternatives have you considered?**
N/A

**Do you have any additional context?**
N/A

### PR DESCRIPTION
**Pull request by amsiglan**
### Description
This PR adds a new monitor type - Composite monitor which provides user with ability to combine existing monitors using trigger conditions and generate alerts when the trigger conditions are satisfied.
 
### Issues Resolved
#610 
 
### Check List
- [] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting-dashboards-plugin/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by lezzago** [comment]
Overall in the PR there are many terms of `Associate` and that is for Composite monitors. This will get confused with feature anywhere, where we associate monitors to visualizations. Should we do a term change here where we use only `Delegate Monitors` and `Set Delegate Monitors`?

File: public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+    setFormikValues(selected, monitorIdx, form);
+    updateSelection(selected);
+  };
+
+  const isValid = () => Object.keys(selection).length > 1;
+
+  const validate = () => {
+    if (!isValid()) return 'Required.';
+  };
```
Comment (lezzago) [inline public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js:177]: We can use the [required helper validate function](https://github.com/opensearch-project/alerting-dashboards-plugin/blob/main/public/utils/validate.js#L60)

File: public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                  <EuiFlexItem
+                    grow={false}
+                    style={{
+                      margin: 0,
+                      padding: 0,
+                      width: '20px',
+                      height: '20px',
+                      position: 'absolute',
+                      right: '-50px',
+                      bottom: '24px',
```
Comment (lezzago) [inline public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js:254]: Does this work well with different scaling screens?

File: public/pages/CreateMonitor/containers/CreateMonitor/utils/constants.js [hunk 3]
```diff
@@ -23,6 +22,17 @@ export const FORMIK_INITIAL_WHERE_EXPRESSION_VALUES = {
   fieldRangeEnd: undefined,
 };
 
+export const DEFAULT_ASSOCIATED_MONITORS_VALUE = {
+  sequence: {
+    delegates: [
+      /*{
+        order: 1,
+        monitor_id: '{{m1}}',
```
Comment (lezzago) [inline public/pages/CreateMonitor/containers/CreateMonitor/utils/constants.js:30]: Is this intentional or dev code?

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/CompositeTriggerCondition.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+}) => {
+  const formikFullFieldName = `${formikFieldPath}${formikFieldName}`;
+  const [graphUi, setGraphUi] = useState(values.searchType === 'graph');
+
+  useEffect(() => {
+    setGraphUi(values.searchType === 'graph');
+  }, [values.searchType]);
+
+  const isValid = () => true;
+  const validate = () => {};
```
Comment (lezzago) [inline public/pages/CreateTrigger/components/CompositeTriggerCondition/CompositeTriggerCondition.js:26]: Why is this needed?

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/CompositeTriggerCondition.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  triggerIndex,
+}) => {
+  const formikFullFieldName = `${formikFieldPath}${formikFieldName}`;
+  const [graphUi, setGraphUi] = useState(values.searchType === 'graph');
+
+  useEffect(() => {
+    setGraphUi(values.searchType === 'graph');
+  }, [values.searchType]);
+
+  const isValid = () => true;
```
Comment (lezzago) [inline public/pages/CreateTrigger/components/CompositeTriggerCondition/CompositeTriggerCondition.js:25]: Why is this needed?

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    'gm'
+  );
+  const matcher = condition.matchAll(queryToExpressionRegex);
+  let match;
+  let expressions = [];
+  let counter = 0;
+  while ((match = matcher.next().value)) {
+    if (counter && !match[1]) return []; // Didn't find condition after the first match
+
+    const monitorId = match[3]?.trim(); // match [3] is the monitor_id
```
Comment (lezzago) [inline public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js:37]: Is it guaranteed to always have 4 values in `matcher.next().value`? If not, should we do stuff like `match?.[3]`?

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionEditor.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+  const isInvalid = (name, form) => !form.values[name]?.length;
+
+  const hasError = (name, form) => {
+    return !form.values[name]?.length && 'Invalid condition.';
+  };
+
+  const validate = (value) => {
+    if (!value?.length) return 'Invalid condition.';
+  };
```
Comment (lezzago) [inline public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionEditor.js:34]: Can you use these here?
https://github.com/opensearch-project/alerting-dashboards-plugin/blob/main/public/utils/validate.js#L13-L18

File: server/services/AlertService.js [hunk 8]
```diff
@@ -28,8 +29,12 @@ export default class AlertService {
       severityLevel = 'ALL',
       alertState = 'ALL',
       monitorIds = [],
+      monitorType = 'monitor',
     } = req.query;
 
+    console.log('get alerts req query from frontend');
+    console.log(JSON.stringify(req.query));
```
Comment (lezzago) [inline server/services/AlertService.js:36]: remove dev code.

File: server/services/AlertService.js [hunk 9]
```diff
@@ -72,20 +77,33 @@ export default class AlertService {
...
+    if (monitorIds.length > 0) {
+      const idField = monitorType === 'composite' ? 'workflowIds' : 'monitorId';
+      params[idField] = !Array.isArray(monitorIds) ? monitorIds : monitorIds[0];
+    }
 
     const { callAsCurrentUser } = this.esDriver.asScoped(req);
     try {
+      console.log('Get Alerts params ****');
+      console.log(monitorType);
+      console.log(JSON.stringify(params));
```
Comment (lezzago) [inline server/services/AlertService.js:89]: remove dev code.

File: server/services/AlertService.js [hunk 10]
```diff
@@ -72,20 +77,33 @@ export default class AlertService {
...
+          id,
+          ...alert,
+          version,
+          alert_source: !!alert.workflow_id ? 'workflow' : 'monitor',
+        };
       });
       const totalAlerts = resp.totalAlerts;
 
+      console.log('Get alerts response *****');
+      console.log(JSON.stringify(alerts));
```
Comment (lezzago) [inline server/services/AlertService.js:105]: remove dev code.

File: server/services/AlertService.js [hunk 11]
```diff
@@ -103,4 +121,32 @@ export default class AlertService {
       });
     }
   };
+
+  getWorkflowAlerts = async (context, req, res) => {
+    console.log('****** GET WORKFLOW ALERTS *****');
+    console.log('get alerts req query from frontend');
+    console.log(JSON.stringify(req.query));
```
Comment (lezzago) [inline server/services/AlertService.js:128]: remove dev code.

File: server/services/AlertService.js [hunk 12]
```diff
@@ -103,4 +121,32 @@ export default class AlertService {
...
+
+    const { callAsCurrentUser } = this.esDriver.asScoped(req);
+    try {
+      const resp = await callAsCurrentUser('alerting.getWorkflowAlerts', req.query);
+      console.log('Get workflow alerts response *****');
+      console.log(JSON.stringify(resp));
```
Comment (lezzago) [inline server/services/AlertService.js:134]: remove dev code.

**Comment by lezzago** [comment]
Also we might want to consolidate the terms composite and workflows

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js (hunk 6 above)
Comment (amsiglan) [reply public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js:37]: If the match doesn't have 4 values, we already have the check for the 4th value - `match[3]?.trim()`

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionEditor.js (hunk 7 above)
Comment (amsiglan) [reply public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionEditor.js:34]: Those utilities check for error, whereas these check for value

File: public/pages/CreateTrigger/containers/CreateTrigger/utils/constants.js [hunk 13]
```diff
@@ -67,6 +68,17 @@ export const FORMIK_INITIAL_TRIGGER_VALUES = {
...
 
+export const FORMIK_COMPOSITE_INITIAL_TRIGGER_VALUES = {
+  name: '',
+  severity: '1',
+  script: {
+    lang: 'painless',
+    source: ``,
+  },
+  triggerConditions: '',
+  actions: undefined,
```
[the remark this answers is not in the document]
Comment (amsiglan) [reply public/pages/CreateTrigger/containers/CreateTrigger/utils/constants.js:79]: In our case triggerConditions needs to be a string, since that is the final type used to send the payload.

File: public/pages/Dashboard/containers/Dashboard.js [hunk 14]
```diff
@@ -166,7 +164,9 @@ export default class Dashboard extends Component {
     const { httpClient } = this.props;
     const { alertsByTriggers } = this.state;
     this.setState({ ...this.state, loadingMonitors: true });
-    const monitorIds = alertsByTriggers.map((alert) => alert.monitor_id);
+    const monitorIds = Array.from(
+      new Set(alertsByTriggers.map((alert) => alert.monitor_id).filter((monitorId) => !!monitorId))
```
[the remark this answers is not in the document]
Comment (amsiglan) [reply public/pages/Dashboard/containers/Dashboard.js:168]: Yes, filtering out ones that are empty. Otherwise backend throws error

File: public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js (hunk 2 above)
Comment (amsiglan) [reply public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js:254]: Let me fix this in follow up PR

File: public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js (hunk 1 above)
Comment (amsiglan) [reply public/pages/CreateMonitor/components/AssociateMonitors/components/MonitorsList.js]: Done

File: public/pages/CreateMonitor/containers/CreateMonitor/utils/constants.js (hunk 3 above)
Comment (amsiglan) [reply public/pages/CreateMonitor/containers/CreateMonitor/utils/constants.js]: That was added as a sample delegate, moving it to JS Doc.

File: public/pages/CreateMonitor/containers/WorkflowDetails/WorkflowDetails.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+import React, { Fragment } from 'react';
+import ContentPanel from '../../../../components/ContentPanel';
+import Schedule from '../../components/Schedule';
+import AssociateMonitors from '../../components/AssociateMonitors/AssociateMonitors';
+import { EuiSpacer } from '@elastic/eui';
+import { MONITOR_TYPE, SEARCH_TYPE } from '../../../../utils/constants';
+
+const WorkflowDetails = ({ values, isDarkMode, httpClient, errors }) => {
+  const isAd = values.searchType === SEARCH_TYPE.AD;
+  const isComposite = values.monitor_type === MONITOR_TYPE.COMPOSITE_LEVEL;
```
[the remark this answers is not in the document]
Comment (amsiglan) [reply public/pages/CreateMonitor/containers/WorkflowDetails/WorkflowDetails.js]: Initially this check wasn't happening in the parent component, but I changed that and didn't update this. Changed it.

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/CompositeTriggerCondition.js (hunk 5 above)
Comment (amsiglan) [reply public/pages/CreateTrigger/components/CompositeTriggerCondition/CompositeTriggerCondition.js]: Same as above

File: public/pages/Dashboard/containers/Dashboard.js [hunk 16]
```diff
@@ -192,26 +192,25 @@ export default class Dashboard extends Component {
...
-      const { id, monitor_id: monitorId } = alert;
-      if (monitorAlerts[monitorId]) monitorAlerts[monitorId].push(id);
-      else monitorAlerts[monitorId] = [id];
+      const id = alert.id;
+      const monitorId = alert.workflow_id || alert.monitor_id;
+      if (monitorAlerts[monitorId]) monitorAlerts[monitorId].alerts.push(id);
+      else
+        monitorAlerts[monitorId] = {
+          alerts: [id],
+          poolType: !!alert.workflow_id ? 'workflows' : 'monitors',
```
Comment (AWSHurneyt) [inline public/pages/Dashboard/containers/Dashboard.js:206]: Non-blocking nitpick: Should we have a constant that contains an exhaustive list of the supported `poolType`s?

File: cypress/integration/composite_level_monitor_spec.js [hunk 17]
```diff
@@ -55,6 +55,7 @@ describe('CompositeLevelMonitor', () => {
 
     it('by visual editor', () => {
       // Select visual editor for method of definition
+      // cy.intercept('GET', '/api/notifications/get_configs?*', channelResponse);
```
Comment (lezzago) [inline cypress/integration/composite_level_monitor_spec.js:58]: remove commented code

File: public/pages/Dashboard/utils/helpers.js [hunk 18]
```diff
@@ -45,8 +48,9 @@ export function addFirstAlert(firstAlert) {
     severity,
     start_time,
     last_notification_time,
-    monitor_name,
-    monitor_id,
+    monitor_name: monitor_name || workflow_name,
+    monitor_id: monitor_id || workflow_id,
```
Comment (AWSHurneyt) [inline public/pages/Dashboard/utils/helpers.js:52]: Would this influence the `undefined` `monitor_id` check on [line 168 of Dashboard.js](https://github.com/opensearch-project/alerting-dashboards-plugin/pull/611/files#diff-90a4d66c6d6bea36116eea37f5ab884f0bb10fce8f7bc1997fc7c24cb01a9f05R168).

File: public/pages/MonitorDetails/containers/Triggers/Triggers.js [hunk 19]
```diff
@@ -24,12 +25,35 @@ export function getUnwrappedTriggers(monitor) {
...
+      case MONITOR_TYPE.COMPOSITE_LEVEL:
+        return trigger[TRIGGER_TYPE.COMPOSITE_LEVEL];
       default:
         return trigger[TRIGGER_TYPE.QUERY_LEVEL];
     }
   });
 }
 
+const expressionsToCondition = (expressions) => {
+  const conditionMap = {
```
Comment (AWSHurneyt) [inline public/pages/MonitorDetails/containers/Triggers/Triggers.js:37]: Non-blocking nitpick: I think I've seen this map in a few places. Could be worthwhile to extract it into a helper file somewhere.

File: public/pages/MonitorDetails/containers/Triggers/Triggers.js [hunk 20]
```diff
@@ -39,12 +63,31 @@ export default class Triggers extends Component {
...
+    if (monitor.monitor_type === MONITOR_TYPE.COMPOSITE_LEVEL) {
+      for (const trigger of triggers) {
+        const { condition, id } = trigger;
+        await this.initializeTriggerConditionsWithMonitorNames(condition.script.source, id);
+      }
+
+      this.setState({ items: triggers });
+    } else {
+      this.setState({ items: triggers });
+    }
```
Comment (AWSHurneyt) [inline public/pages/MonitorDetails/containers/Triggers/Triggers.js:100]: Non-blocking nitpick: It looks like `this.setState({ items: triggers });` is called in both the `if/else` blocks; so it could be simplified by just moving it out of the block.

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js (hunk 6 above)
Comment (lezzago) [reply public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js:37]: Yea but if the index `3` is not there, wont it throw array out of bounds exception since we first do `match[3]` and then do the trim call.

File: public/utils/helpers.js [hunk 21]
```diff
@@ -47,6 +47,22 @@ export const inputLimitText = (
   );
 };
 
+export async function deleteMonitor(monitor, httpClient, notifications) {
+  const { id, version } = monitor;
+  const poolType = monitor.item_type === 'composite' ? 'workflows' : 'monitors';
```
Comment (AWSHurneyt) [inline public/utils/helpers.js:52]: Non-blocking nitpick: It may be worthwhile to compile an exhaustive list of `item_type`s in a constant somewhere.

File: public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js (hunk 6 above)
Comment (amsiglan) [reply public/pages/CreateTrigger/components/CompositeTriggerCondition/ExpressionBuilder.js:37]: No, in JS it would simply return `undefined` and not throw exception

File: server/services/MonitorService.js [hunk 22]
```diff
@@ -137,7 +238,10 @@ export default class MonitorService {
       }
 
       const { callAsCurrentUser } = await this.esDriver.asScoped(req);
-      const updateResponse = await callAsCurrentUser('alerting.updateMonitor', params);
+      const updateResponse = await callAsCurrentUser(
+        `alerting.${type === 'workflow' ? 'updateWorkflow' : 'updateMonitor'}`,
```
Comment (AWSHurneyt) [inline server/services/MonitorService.js:242]: Non-blocking nitpick: It may be worthwhile to compile an exhaustive list of `type`s in a constant somewhere.

**Review by lezzago — APPROVED** [review]

**Review by AWSHurneyt — COMMENTED** [review]
Most of my comments are non-blocking; so I think it would be fine to create a github issue to track any of those changes for a future release.

File: cypress/fixtures/sample_composite_level_monitor.json [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+          "sequence": {
+            "delegates": [
+              {
+                "order": 1,
+                "monitor_id": "qdYBw4gB2qeAWe54jQyZ"
+              },
+              {
+                "order": 2,
+                "monitor_id": "rtYBw4gB2qeAWe54wAx5"
+              }
```
[the remark this answers is not in the document]
Comment (amsiglan) [reply cypress/fixtures/sample_composite_level_monitor.json:27]: These monitor Ids are for existing monitors which we are specifying in the creation payload for a composite monitor

File: cypress/integration/composite_level_monitor_spec.js [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+  describe('can be created', () => {
+    beforeEach(() => {
+      // Visit Alerting OpenSearch Dashboards
+      cy.visit(`${Cypress.env('opensearch_dashboards')}/app/${PLUGIN_NAME}#/monitors`);
+
+      // Common text to wait for to confirm page loaded, give up to 20 seconds for initial load
+      cy.contains('Create monitor', { timeout: 20000 });
+
+      // Go to create monitor page
+      cy.contains('Create monitor').click({ force: true });
```
[the remark this answers is not in the document]
Comment (amsiglan) [reply cypress/integration/composite_level_monitor_spec.js:50]: Let me do it in follow up PR

File: public/pages/CreateMonitor/components/AssociateMonitors/AssociateMonitors.js [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+      <EuiText color={'subdued'} size={'xs'}>
+        Delegate two or more monitors to run as part of this workflow. The monitor types per query,
+        per bucket, and per document are supported.{' '}
+        <EuiLink
+          href={
+            'https://opensearch.org/docs/latest/observing-your-data/alerting/monitors/#monitor-types'
+          }
+          target="_blank"
+        >
+          Learn more.
```
[the remark this answers is not in the document]
Comment (amsiglan) [reply public/pages/CreateMonitor/components/AssociateMonitors/AssociateMonitors.js:60]: Let me fix this in a follow up PR

File: public/pages/CreateMonitor/components/MonitorType/MonitorType.js [hunk 26]
```diff
@@ -38,27 +41,36 @@ const onChangeDefinition = (e, form) => {
...
-    bucket.
+    Per bucket monitors run a query that evaluates trigger criteria based on aggregated values in
+    the dataset.
   </EuiText>
 );
 
 const clusterMetricsDescription = (
   <EuiText color={'subdued'} size={'xs'} style={{ paddingBottom: '10px', paddingTop: '0px' }}>
-    Per cluster metrics monitors allow you to alert based on responses to common REST APIs.
+    Per cluster metrics monitors run API requests to monitor the cluster’s health.
```
[the remark this answers is not in the document]
Comment (amsiglan) [reply public/pages/CreateMonitor/components/MonitorType/MonitorType.js:58]: The text came from content writers, so let's evaluate this as a fast follow

**Review by AWSHurneyt — APPROVED** [review]

File: cypress/fixtures/sample_composite_level_monitor.json (hunk 23 above)
[the remark this answers is not in the document]
Comment (AWSHurneyt) [reply cypress/fixtures/sample_composite_level_monitor.json:27]: Ah, I didn't realize we now had the capability to provide the monitor ID when creating a monitor; that must be new. Good to know.

File: public/pages/Dashboard/utils/helpers.js (hunk 18 above)
Comment (AWSHurneyt) [reply public/pages/Dashboard/utils/helpers.js:52]: Discussed offline. @amsiglan explained that it's fine for the monitorID to be either the `monitor_id` or the `workflow_id`; we just don't want the monitorId to be blank.

**Comment by amsiglan** [comment]
All nitpicks will be addressed in follow up PR
